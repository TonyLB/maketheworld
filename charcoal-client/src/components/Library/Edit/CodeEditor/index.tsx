import { FunctionComponent, useCallback, useMemo, useState } from "react";
import { createEditor, Descendant } from "slate";
import { withHistory } from "slate-history";
import { Editable, Slate, withReact } from "slate-react";
import stringToSlate, { slateToString } from "./stringToSlate";

type CodeEditorProps = {
    source: string;
    onChange: (value: string) => void;
}

const Leaf = ({ attributes, children, leaf }: { attributes: any, children: any, leaf: any }) => {
    return (
        <span
            {...attributes}
            style={{ backgroundColor: leaf.error && '#ffeeba' }}
            spellCheck={false}
        >
            {children}
        </span>
    )
}

export const CodeEditor: FunctionComponent<CodeEditorProps> = ({ source, onChange }) => {
    const [editor] = useState(() => withHistory(withReact(createEditor())))
    const value = useMemo(() => (stringToSlate(source)), [source])
    const onChangeHandler = useCallback((nodes: Descendant[]) => {
        onChange(slateToString(nodes))
    }, [onChange])
    const renderLeaf = useCallback(props => (<Leaf { ...props } />), [])
    return <Slate editor={editor} value={value} onChange={onChangeHandler}>
        <Editable
            renderLeaf={renderLeaf}
        />
    </Slate>
}

export default CodeEditor
