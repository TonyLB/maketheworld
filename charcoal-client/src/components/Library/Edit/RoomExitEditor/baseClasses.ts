export type RoomExit = {
    key: string;
    to: string;
    from: string;
    name: string;
}
