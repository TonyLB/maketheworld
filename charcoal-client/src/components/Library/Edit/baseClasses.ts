import { BaseEditor, Selection } from 'slate'
import { ReactEditor } from 'slate-react'

export type CustomText = {
    text: string;
}

export type CustomLineElement = {
    type: 'line';
    children: CustomText[]
}

export type CustomActionLinkElement = {
    type: 'actionLink';
    to: string;
    key: string;
    children: CustomText[]
}

export type CustomFeatureLinkElement = {
    type: 'featureLink';
    to: string;
    key: string;
    children: CustomText[]
}

export type CustomLinkElement = CustomActionLinkElement | CustomFeatureLinkElement

export type CustomLineBreak = {
    type: 'lineBreak';
}

export type CustomParagraphContents = CustomText | CustomActionLinkElement | CustomFeatureLinkElement | CustomLineBreak

export const isCustomLineBreak = (item: CustomParagraphContents): item is CustomLineBreak => ('type' in item && item.type === 'lineBreak')
export const isCustomActionLink = (item: CustomParagraphContents): item is CustomActionLinkElement => ('type' in item && item.type === 'actionLink')
export const isCustomFeatureLink = (item: CustomParagraphContents): item is CustomFeatureLinkElement => ('type' in item && item.type === 'featureLink')
export const isCustomLink = (item: CustomParagraphContents): item is CustomLinkElement => (isCustomActionLink(item) || isCustomFeatureLink(item))
export const isCustomText = (item: CustomParagraphContents): item is CustomText => ('text' in item)

export type CustomParagraphElement = {
    type: 'paragraph';
    children: CustomParagraphContents[]
}

export type CustomDescriptionElement = {
    type: 'description';
    children: CustomParagraphElement[]
}

type CustomElement = CustomLineElement |
    CustomActionLinkElement |
    CustomFeatureLinkElement |
    CustomParagraphElement |
    CustomDescriptionElement



declare module 'slate' {
    interface CustomTypes {
      Editor: BaseEditor & ReactEditor & { saveSelection?: Selection }
      Element: CustomElement
      Text: CustomText
    }
  }

export {}