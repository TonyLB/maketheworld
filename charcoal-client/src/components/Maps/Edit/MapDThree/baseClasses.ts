import {
    SimulationNodeDatum,
    SimulationLinkDatum
} from 'd3-force'

export type SimNode = SimulationNodeDatum & {
    id: string;
    zLevel?: number;
    cascadeNode: boolean;
    roomId: string;
    visible: boolean;
}
export type NodeRecord = Record<string, SimNode>
export type LinkRecord = {
    id: string,
    source: string,
    target: string,
    visible?: boolean
}[]
export type SimulationReturn = {
    key: string,
    nodes: SimNode[],
    links: SimulationLinkDatum<SimNode>[]
}

export type SimCallback = (nodes: SimNode[]) => void

export type MapNodes = SimNode[]
export type MapLinks = SimulationLinkDatum<SimulationNodeDatum>[]
