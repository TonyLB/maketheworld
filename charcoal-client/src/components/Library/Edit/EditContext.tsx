import React, { FunctionComponent, useContext } from "react"
import { GenericTreeNode } from "@tonylb/mtw-wml/dist/tree/baseClasses";
import { SchemaTag } from "@tonylb/mtw-wml/dist/schema/baseClasses";
import { TreeId } from "@tonylb/mtw-wml/dist/tree/baseClasses";

type EditContextType = {
    field: GenericTreeNode<SchemaTag, TreeId>;
    parentId: string;
    tag: 'ShortName' | 'Name' | 'Description' | 'Statement' | 'Fallthrough' | 'If'
}

const EditContext = React.createContext<EditContextType>({
    field: { data: { tag: 'Name' }, id: '', children: [] },
    parentId: '',
    tag: 'Name'
})

export const EditSchema: FunctionComponent<EditContextType> = ({ field, parentId, tag, children }) => {
    return <EditContext.Provider value={{ field, parentId, tag }}>
        { children }
    </EditContext.Provider>
}

export const useEditContext = () => (useContext(EditContext))
