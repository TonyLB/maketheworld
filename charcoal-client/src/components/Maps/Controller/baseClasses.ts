import { GenericTree } from "@tonylb/mtw-sequence/dist/tree/baseClasses"
import { SchemaConditionTag, SchemaExitTag, SchemaRoomTag } from "@tonylb/mtw-wml/dist/simpleSchema/baseClasses"

export type ToolSelected = 'Select' | 'Move' | 'AddRoom' | 'OneWayExit' | 'TwoWayExit'

export type MapTreeExit = SchemaExitTag & { inherited?: boolean }
export type MapTreeRoom = SchemaRoomTag & { inherited?: boolean }
export type MapTreeCondition = SchemaConditionTag & { inherited?: boolean }

export type MapTreeItem = MapTreeExit | MapTreeRoom | MapTreeCondition

type MapContextExitDrag = {
    sourceRoomId: string;
    x: number;
    y: number;
}

type MapDispatchSetTool = {
    type: 'SetToolSelected';
    value: ToolSelected;
}

type MapDispatchSetExitDrag = {
    type: 'SetExitDrag';
} & Partial<MapContextExitDrag>

export type MapDispatchAction = MapDispatchSetTool | MapDispatchSetExitDrag

export type MapContextType = {
    mapId: string;
    tree: GenericTree<MapTreeItem>;
    UI: {
        //
        // The Map editor can conceivably need data for:
        //    - Which tool is selected in the toolbar
        //    - Whether an exit is being dragged, from where, and to where
        // Updates to this data should be performed through the mapDispatch
        // function.
        //
        toolSelected: ToolSelected;
        exitDrag: MapContextExitDrag;
    },
    // mapD3: MapDThree,
    mapDispatch: (action: MapDispatchAction) => void;
}
