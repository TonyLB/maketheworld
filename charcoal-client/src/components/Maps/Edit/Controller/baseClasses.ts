import { SchemaConditionTag, SchemaExitTag, SchemaRoomTag } from "@tonylb/mtw-wml/dist/simpleSchema/baseClasses"

export type ToolSelected = 'Select' | 'Move' | 'AddRoom' | 'OneWayExit' | 'TwoWayExit'

export type MapTreeExit = SchemaExitTag & { inherited?: boolean }
export type MapTreeRoom = SchemaRoomTag & { inherited?: boolean; contents: MapTreeItem[] }
export type MapTreeCondition = SchemaConditionTag & { inherited?: boolean; contents: MapTreeItem[] }

export type MapTreeItem = MapTreeExit | MapTreeRoom | MapTreeCondition
