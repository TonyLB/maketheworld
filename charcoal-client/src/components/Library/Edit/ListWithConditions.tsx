import React, { FunctionComponent, PropsWithChildren, ReactElement } from "react"

import List from "@mui/material/List"
import ListItem from "@mui/material/ListItem"
import ListItemIcon from "@mui/material/ListItemIcon"
import ListItemButton from "@mui/material/ListItemButton"
import ListItemText from "@mui/material/ListItemText"
import AddIcon from '@mui/icons-material/Add'

import { isSchemaCondition, SchemaConditionTag, SchemaTag } from "@tonylb/mtw-wml/dist/schema/baseClasses"
import { GenericTreeFiltered, GenericTreeNodeFiltered, treeNodeTypeguard } from "@tonylb/mtw-wml/dist/tree/baseClasses"
import { EditSubListSchema, useEditContext } from "./EditContext"
import IfElseTree from "./IfElseTree"
import { maybeGenericIDFromTree } from "@tonylb/mtw-wml/dist/tree/genericIDTree"

type ListWithConditionsProperties<T extends SchemaTag> = {
    typeGuard: (value: SchemaTag) => value is T;
    render: FunctionComponent<{ item: GenericTreeNodeFiltered<T, SchemaTag>, index: number }>;
    label: string;
    defaultNode: T;
}

type ListWithConditionsGroup<T extends SchemaTag> = {
    type: 'tree';
    startIndex: number;
    tree: GenericTreeFiltered<T, SchemaTag>;
} | {
    type: 'if';
    index: number;
    node: GenericTreeNodeFiltered<SchemaConditionTag, SchemaTag>;
}

export const ListWithConditions = <T extends SchemaTag>(props: PropsWithChildren<ListWithConditionsProperties<T>>): ReactElement<any, any> | null => {
    const { render: Render } = props
    const { value, onChange } = useEditContext()

    const listGroups = value.reduce<ListWithConditionsGroup<T>[]>((previous, item, index) => {
        if (treeNodeTypeguard(isSchemaCondition)(item)) {
            return [...previous, { type: 'if', node: item, index }]
        }
        else if (treeNodeTypeguard(props.typeGuard)(item)) {
            const last = previous.length ? previous.slice(-1)[0] : undefined
            if (last && last.type === 'tree') {
                return [...previous.slice(0, -1), { ...last, tree: [...last.tree, item] }]
            }
            else {
                return [...previous, { type: 'tree', tree: [item], startIndex: index }]
            }
        }
        else {
            throw new Error('Invalid type in ListWithConditions')
        }
    }, [])

    const addButton = <List>
        <ListItemButton onClick={() => { onChange(maybeGenericIDFromTree([...value, { data: props.defaultNode, children: [] }])) }}>
            <ListItemIcon>
                <AddIcon />
            </ListItemIcon>
            <ListItemText primary={`Add ${props.label}`} />
        </ListItemButton>
    </List>

    return <React.Fragment>
        { listGroups.map((group, index) => (
            group.type === 'tree'
                ? <List key={`list-with-conditions-${group.startIndex + index}`}>
                    { group.tree.map((item, index) => (
                        <EditSubListSchema
                            index={group.startIndex + index}
                        >
                            <ListItem><Render item={item} index={group.startIndex + index} /></ListItem>
                        </EditSubListSchema>
                    ))}
                    { index === listGroups.length - 1 && addButton }
                </List>
                : <EditSubListSchema
                    key={`list-with-conditions-${group.index}`}
                    index={group.index}
                >
                    <IfElseTree render={props.render} />
                </EditSubListSchema>
        )) }
        { listGroups.slice(-1)[0]?.type !== 'tree' && addButton }
    </React.Fragment>
}

export default ListWithConditions
