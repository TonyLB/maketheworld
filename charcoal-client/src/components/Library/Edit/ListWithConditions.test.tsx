jest.mock('./CodeEditor')
import CodeEditor from './CodeEditor'

import renderer from 'react-test-renderer'
import React, { FunctionComponent } from 'react'

import ListWithConditions from './ListWithConditions'
import { isSchemaExit, isSchemaOutputTag } from '@tonylb/mtw-wml/dist/schema/baseClasses'
import { treeNodeTypeguard } from '@tonylb/mtw-wml/dist/tree/baseClasses'
import { schemaOutputToString } from '@tonylb/mtw-wml/dist/schema/utils/schemaOutput/schemaOutputToString'
import { EditSchema, useEditContext } from './EditContext'
import { treeTypeGuard } from '@tonylb/mtw-wml/dist/tree/filter'

describe('ListWithConditions component', () => {
    const render: FunctionComponent<{}> = () => {
        const { value } = useEditContext()
        return <React.Fragment>
            { value
                .filter(treeNodeTypeguard(isSchemaExit))
                .map(({ data, children }) => (<div>{ `'${data.from}' to '${data.to}': ${schemaOutputToString(treeTypeGuard({ typeGuard: isSchemaOutputTag, tree: children }))}` }</div>))
            }
        </React.Fragment>
    }

    beforeEach(() => {
        (CodeEditor as jest.Mock).mockReturnValue(null)
    })

    it('renders non-conditions correctly', () => {
        expect(renderer
            .create(
                <EditSchema
                    value={[
                        { data: { tag: 'Exit', from: 'room1', to: 'room2', key: 'room1#room2' }, children: [{ data: { tag: 'String', value: 'closet' }, children: [] }]},
                        { data: { tag: 'Exit', from: 'room1', to: 'room3', key: 'room1#room3' }, children: [{ data: { tag: 'String', value: 'lobby' }, children: [] }]}
                    ]}
                    onChange={() => {}}
                >
                    <ListWithConditions
                        typeGuard={isSchemaExit}
                        render={render}
                    />
                </EditSchema>
            ).toJSON()
        ).toMatchSnapshot()
    })

    it('renders conditions correctly', () => {
        expect(renderer
            .create(
                <EditSchema
                    value={[
                        { data: { tag: 'Exit', from: 'room1', to: 'room2', key: 'room1#room2' }, children: [{ data: { tag: 'String', value: 'closet' }, children: [] }]},
                        { data: { tag: 'If' }, children: [
                            { data: { tag: 'Statement', if: 'true' }, children: [
                                { data: { tag: 'Exit', from: 'room1', to: 'room3', key: 'room1#room3' }, children: [{ data: { tag: 'String', value: 'lobby' }, children: [] }]}
                            ] }
                        ] }
                    ]}
                    onChange={() => {}}
                >
                    <ListWithConditions
                        typeGuard={isSchemaExit}
                        render={render}
                    />
                </EditSchema>
            ).toJSON()
        ).toMatchSnapshot()
    })

})