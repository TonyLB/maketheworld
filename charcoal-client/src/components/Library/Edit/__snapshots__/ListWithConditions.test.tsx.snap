// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ListWithConditions component renders conditions correctly 1`] = `
Array [
  <ul
    className="MuiList-root MuiList-padding css-h4y409-MuiList-root"
  >
    <li
      className="MuiListItem-root MuiListItem-gutters MuiListItem-padding css-1p823my-MuiListItem-root"
      disabled={false}
    >
      <div>
        'room1' to 'room2': closet
      </div>
    </li>
  </ul>,
  <div
    className="MuiBox-root css-n8b9os"
  >
    <div
      className="MuiBox-root css-15r38cd"
    >
      If
      <div
        className="MuiBox-root css-1jbq9h"
      />
    </div>
    <div
      className="MuiBox-root css-90szgu"
      onClick={[Function]}
    >
      <span>
        <div>
          'room1' to 'room3': lobby
        </div>
      </span>
    </div>
    <div
      className="MuiBox-root css-1lzl05c"
    >
      <div
        aria-disabled={true}
        className="MuiButtonBase-root Mui-disabled MuiChip-root MuiChip-filled Mui-disabled MuiChip-sizeSmall MuiChip-colorDefault MuiChip-clickable MuiChip-clickableColorDefault MuiChip-filledDefault css-13cuktm-MuiButtonBase-root-MuiChip-root"
        onBlur={[Function]}
        onClick={[Function]}
        onContextMenu={[Function]}
        onDragLeave={[Function]}
        onFocus={[Function]}
        onKeyDown={[Function]}
        onKeyUp={[Function]}
        onMouseDown={[Function]}
        onMouseLeave={[Function]}
        onMouseUp={[Function]}
        onTouchEnd={[Function]}
        onTouchMove={[Function]}
        onTouchStart={[Function]}
        role="button"
        tabIndex={-1}
      >
        <svg
          aria-hidden={true}
          className="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium MuiChip-icon MuiChip-iconSmall MuiChip-iconColorDefault css-i4bv87-MuiSvgIcon-root"
          data-testid="AddIcon"
          focusable="false"
          viewBox="0 0 24 24"
        >
          <path
            d="M19 13h-6v6h-2v-6H5v-2h6V5h2v6h6v2z"
          />
        </svg>
        <span
          className="MuiChip-label MuiChip-labelSmall css-wjsjww-MuiChip-label"
        >
          elseIf
        </span>
      </div>
      <div
        aria-disabled={true}
        className="MuiButtonBase-root Mui-disabled MuiChip-root MuiChip-filled Mui-disabled MuiChip-sizeSmall MuiChip-colorDefault MuiChip-clickable MuiChip-clickableColorDefault MuiChip-filledDefault css-13cuktm-MuiButtonBase-root-MuiChip-root"
        onBlur={[Function]}
        onClick={[Function]}
        onContextMenu={[Function]}
        onDragLeave={[Function]}
        onFocus={[Function]}
        onKeyDown={[Function]}
        onKeyUp={[Function]}
        onMouseDown={[Function]}
        onMouseLeave={[Function]}
        onMouseUp={[Function]}
        onTouchEnd={[Function]}
        onTouchMove={[Function]}
        onTouchStart={[Function]}
        role="button"
        tabIndex={-1}
      >
        <svg
          aria-hidden={true}
          className="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium MuiChip-icon MuiChip-iconSmall MuiChip-iconColorDefault css-i4bv87-MuiSvgIcon-root"
          data-testid="AddIcon"
          focusable="false"
          viewBox="0 0 24 24"
        >
          <path
            d="M19 13h-6v6h-2v-6H5v-2h6V5h2v6h6v2z"
          />
        </svg>
        <span
          className="MuiChip-label MuiChip-labelSmall css-wjsjww-MuiChip-label"
        >
          else
        </span>
      </div>
    </div>
    <div
      className="MuiBox-root css-klha72"
    >
      <button
        className="MuiButtonBase-root MuiIconButton-root MuiIconButton-sizeSmall css-1pe4mpk-MuiButtonBase-root-MuiIconButton-root"
        disabled={false}
        onBlur={[Function]}
        onClick={[Function]}
        onContextMenu={[Function]}
        onDragLeave={[Function]}
        onFocus={[Function]}
        onKeyDown={[Function]}
        onKeyUp={[Function]}
        onMouseDown={[Function]}
        onMouseLeave={[Function]}
        onMouseUp={[Function]}
        onTouchEnd={[Function]}
        onTouchMove={[Function]}
        onTouchStart={[Function]}
        tabIndex={0}
        type="button"
      >
        <svg
          aria-hidden={true}
          className="MuiSvgIcon-root MuiSvgIcon-fontSizeSmall css-ptiqhd-MuiSvgIcon-root"
          data-testid="DeleteIcon"
          focusable="false"
          viewBox="0 0 24 24"
        >
          <path
            d="M6 19c0 1.1.9 2 2 2h8c1.1 0 2-.9 2-2V7H6v12zM19 4h-3.5l-1-1h-5l-1 1H5v2h14V4z"
          />
        </svg>
        <span
          className="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
        />
      </button>
    </div>
  </div>,
  <ul
    className="MuiList-root MuiList-padding css-h4y409-MuiList-root"
  >
    <div
      className="MuiButtonBase-root MuiListItemButton-root MuiListItemButton-gutters MuiListItemButton-root MuiListItemButton-gutters css-16ac5r2-MuiButtonBase-root-MuiListItemButton-root"
      onBlur={[Function]}
      onClick={[Function]}
      onContextMenu={[Function]}
      onDragLeave={[Function]}
      onFocus={[Function]}
      onKeyDown={[Function]}
      onKeyUp={[Function]}
      onMouseDown={[Function]}
      onMouseLeave={[Function]}
      onMouseUp={[Function]}
      onTouchEnd={[Function]}
      onTouchMove={[Function]}
      onTouchStart={[Function]}
      role="button"
      tabIndex={0}
    >
      <div
        className="MuiListItemIcon-root css-cveggr-MuiListItemIcon-root"
      >
        <svg
          aria-hidden={true}
          className="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
          data-testid="AddIcon"
          focusable="false"
          viewBox="0 0 24 24"
        >
          <path
            d="M19 13h-6v6h-2v-6H5v-2h6V5h2v6h6v2z"
          />
        </svg>
      </div>
      <div
        className="MuiListItemText-root css-tlelie-MuiListItemText-root"
      >
        <span
          className="MuiTypography-root MuiTypography-body1 MuiListItemText-primary css-10hburv-MuiTypography-root"
        >
          Add Exit
        </span>
      </div>
      <span
        className="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
      />
    </div>
  </ul>,
]
`;

exports[`ListWithConditions component renders non-conditions correctly 1`] = `
<ul
  className="MuiList-root MuiList-padding css-h4y409-MuiList-root"
>
  <li
    className="MuiListItem-root MuiListItem-gutters MuiListItem-padding css-1p823my-MuiListItem-root"
    disabled={false}
  >
    <div>
      'room1' to 'room2': closet
    </div>
  </li>
  <li
    className="MuiListItem-root MuiListItem-gutters MuiListItem-padding css-1p823my-MuiListItem-root"
    disabled={false}
  >
    <div>
      'room1' to 'room3': lobby
    </div>
  </li>
  <ul
    className="MuiList-root MuiList-padding css-h4y409-MuiList-root"
  >
    <div
      className="MuiButtonBase-root MuiListItemButton-root MuiListItemButton-gutters MuiListItemButton-root MuiListItemButton-gutters css-16ac5r2-MuiButtonBase-root-MuiListItemButton-root"
      onBlur={[Function]}
      onClick={[Function]}
      onContextMenu={[Function]}
      onDragLeave={[Function]}
      onFocus={[Function]}
      onKeyDown={[Function]}
      onKeyUp={[Function]}
      onMouseDown={[Function]}
      onMouseLeave={[Function]}
      onMouseUp={[Function]}
      onTouchEnd={[Function]}
      onTouchMove={[Function]}
      onTouchStart={[Function]}
      role="button"
      tabIndex={0}
    >
      <div
        className="MuiListItemIcon-root css-cveggr-MuiListItemIcon-root"
      >
        <svg
          aria-hidden={true}
          className="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
          data-testid="AddIcon"
          focusable="false"
          viewBox="0 0 24 24"
        >
          <path
            d="M19 13h-6v6h-2v-6H5v-2h6V5h2v6h6v2z"
          />
        </svg>
      </div>
      <div
        className="MuiListItemText-root css-tlelie-MuiListItemText-root"
      >
        <span
          className="MuiTypography-root MuiTypography-body1 MuiListItemText-primary css-10hburv-MuiTypography-root"
        >
          Add Exit
        </span>
      </div>
    </div>
  </ul>
</ul>
`;
